### ISSUE TITLE
Linter.py: Make it possible to override `executable` for all linter bears

### ISSUE THREAD

**Issue by NiklasMM** [coala/coala#4107]
Say, we have a linter bear called `FooLinterBear`:

``` Python
@linter(executable='foo', output=[...])
class FooLinterBear:
   [...]
```

If a user is in an environment, where they can't make the wrapped tool available as `foo` in the `$PATH` or want to use a different version than is regularly installed, they currently have no way to use the bear, since the executable is hard-coded. I discovered this, when I wanted to use `PyflakesBear` for a Python 2 project: Pyflakes has the same output in both versions, so the linter would work fine, if I could override the executable with the Python 2 version.

I'm imagining something like this:
```ini
[python]
bears=Pyflakesbear
override_executable=/path/to/python2/pyflakes
```

Obviously, this would have to be used with great care and should probably always emit a warning, since it's easy to destroy the functionality of the bear.

I'm not sure if that's something we want to add, so please add your opinions :)

**Comment by Makman2** [comment]
sounds interesting, I definitely see the usecase. Though I don't have any idea how to implement that nicely...
One of the major drawbacks of this (why we definitely should emit a warning) that a coafile might be unusable across multiple machines (because different machines, different exe-paths). That's why we should maybe just encourage people to set up environments properly.

> where they can't make the wrapped tool available as foo in the $PATH

as this should be always possible. Also for your case you can just set `pyflakes` to your python version.

**Comment by jayvdb** [comment]
This could be a nice workaround for https://github.com/coala/coala-bears/issues/929 

And we were recently talking in gitter about running some bears (with complex dependencies) via docker, and a custom command could be a way to prototype that better without modifying core. We could create a fake executable script which invokes docker, but that is more complicated for the user.

**Comment by adtac** [comment]
@NiklasMM so the rest of the arguments won't be changed, yes? We can make that customizable too. But at that point, we're just creating new bears inside the coafile :D

**Comment by adtac** [comment]
One more question: what happens if there are multiple bears in a section? How would I override just one bear's executable?

**Comment by yukiisbored** [comment]
Hmm, what about something like this

```
[all]
additional_path = /path/to/python2

[all.python]
bears=PyflakesBear
override_executable=pyflakes
```

A couple of questions:
* Should we prioritize `additional_path`?
* How do we define it on "per-bear" basis?

**Comment by yukiisbored** [comment]
But yeah, It's a really tough one to swallow because it'll kill portability of coafiles.

**Comment by Makman2** [comment]
The user could set paths in his `.coarc` file, as this one is not shared (usually).

**Comment by li-boxuan** [comment]
Maybe we can have sth like:

```yaml
[all]
additional_path:
  - /path/to/python2
  - /path/to/python3
  - ...
```
coala would check if executable exists on the following paths in order. If the executable is not found, try executable in $PATH

And yeah, it would be good if this setting is put in  `.coarc` file.

**Comment by chay2199** [comment]
After checking if executable exists on the following paths in order and If the executable is not found, trying executable in $PATH, we can initiate a step called 'code freeze' which will make a single file for all executable and IMO this can solve the problem of 'different machines, different exe-paths' as this process will make it independent of path and environment. But this step should be implemented every time there is a change in additional_path. Although I have little knowledge about implementing this right now, but is it a viable option?

**Comment by chay2199** [comment]
@li-boxuan there can be a case where a user wants to switch the executable path again but since we are checking the following paths in order it will create problems. Do you suggest an input method that can take the `path` from user and can reorder the `additional_path` in the `.coarc` file or do you want the user to manually change the `.coarc` file everytime?

**Comment by li-boxuan** [comment]
@chay2199 IMO we could have something that is simple but works first, and then see how we can improve it or handle more complicated cases.

**Comment by chay2199** [comment]
@li-boxuan right now we simply pass the linter name and `shutil.which` return the path to an executable which would be run, but now we have a list of `additional_path` which can contain paths related to more than one bear, how do you propose to check if the path in the list belongs to that particular linter bear? One way is to check the linter's name in the path, for ex /path/to/python2/`pyflakes` but this will not be a generalized way as the user can also use a path which doesn't contain the linter name.

**Comment by li-boxuan** [comment]
> One way is to check the linter's name in the path, for ex /path/to/python2/pyflakes but this will not be a generalized way as the user can also use a path which doesn't contain the linter name.

@chay2199 Yeah, that's a drawback of https://github.com/coala/coala/issues/4107#issuecomment-483748660

IMO that's fine. If users have a different name for executable, they can rename the executable or make a symbolic link. Can it solve the problem proposed by this issue? Yes. Is it backward compatible? Yes. Is it easy to implement? Relatively yes. Is it perfect and flexible? Absolutely no. Is it a good first step to have for this issue? Maybe.

https://github.com/coala/coala/issues/4107#issuecomment-296177511 is able to handle this problem and is a bit more complicated.

**Comment by chay2199** [comment]
> #4107 (comment) is able to handle this problem and is a bit more complicated.
@li-boxuan This will have similar problems once the number of bears increase. But let's just solve this issue first.
